frontend/apps/filemanager/filemanagercollate.lua

```diff
@@ -0,0 +1,171 @@
+                return function(a, b)
+                    if a.opened == b.opened then
+                        if a.opened then
+                            return a.percent_finished < b.percent_finished
+                        end
+                        return ffiUtil.strcoll(a.text, b.text)
+                    end
+                    return b.opened
+                end, cache
+            end,
+            item_func = function(item)
+                local percent_finished
+                item.opened = DocSettings:hasSidecarFile(item.fullpath)
+                if item.opened then
+                    local doc_settings = DocSettings:open(item.fullpath)
+                    percent_finished = doc_settings:readSetting("percent_finished")
+                end
+                item.percent_finished = percent_finished or 0
+                return item
+            end,
+            mandatory_func = function(item)
+                return item.opened and string.format("%d %%", 100 * item.percent_finished) or "–"
+            end,
+        },
+        {
+            id = "percent_unopened_last",
+            text = _("percent - unopened last"),
+            can_collate_mixed = false,
+            init_sort_func = function(cache)
+                return function(a, b)
+                    if a.opened == b.opened then
+                        if a.opened then
+                            return a.percent_finished < b.percent_finished
+                        end
+                        return ffiUtil.strcoll(a.text, b.text)
+                    end
+                    return a.opened
+                end, cache
+            end,
+            item_func = function(item)
+                local percent_finished
+                item.opened = DocSettings:hasSidecarFile(item.fullpath)
+                if item.opened then
+                    local doc_settings = DocSettings:open(item.fullpath)
+                    percent_finished = doc_settings:readSetting("percent_finished")
+                end
+                item.percent_finished = percent_finished or 0
+                return item
+            end,
+            mandatory_func = function(item)
+                return item.opened and string.format("%d %%", 100 * item.percent_finished) or "–"
+            end,
+        },
+    }
+end
+
+function FileManagerCollate:getCollateById(id)
+    if self.collate_by_id == nil then
+        self.collate_by_id = {}
+        for i, c in ipairs(self.getCollates()) do
+            self.collate_by_id[c.id] = c
+        end
+    end
+    return self.collate_by_id[id]
+end
+
+function FileManagerCollate:getCurrentCollate()
+    local collate_id = G_reader_settings:readSetting("collate", "strcoll")
+    local collate = self:getCollateById(collate_id)
+    if collate ~= nil then
+        return collate, collate_id
+    else
+        G_reader_settings:saveSetting("collate", "strcoll")
+        return self:getCollateById("strcoll"), "strcoll"
+    end
+end
```

Comment:
I'm also not necessarily a fan of this approach.

Building on my previous comment, just use a hash instead of an array and just call `FileManagerCollate.collate_list[collate_id]`?

Reply:
Lua hashes have no defined order, so using a simple hash would mean that the menu order is random. I didn't really want to implement a separate ordering for the menu. Other than that, a hash would indeed be nicer, so if you have any ideas for the menu order, I'm open to them.

Reply:
Either live with alphabetical order (via util.orderedPairs), or, yeah, whip something together with an extra order field.